src/hotspot/share/gc/shenandoah/shenandoahFreeSet.cpp

```diff
@@ -32,36 +33,373 @@
+          "rightmost region should be free: " SIZE_FORMAT, rightmost(Mutator));
+
+  // If Mutator set is empty, leftmosts will both equal max, rightmosts will both equal zero.  Likewise for empty region sets.
+  size_t beg_off = leftmosts[Mutator];
+  size_t end_off = rightmosts[Mutator];
+  assert (beg_off >= leftmost(Mutator),
+          "free regions before the leftmost: " SIZE_FORMAT ", bound " SIZE_FORMAT, beg_off, leftmost(Mutator));
+  assert (end_off <= rightmost(Mutator),
+          "free regions past the rightmost: " SIZE_FORMAT ", bound " SIZE_FORMAT,  end_off, rightmost(Mutator));
+
+  beg_off = empty_leftmosts[Mutator];
+  end_off = empty_rightmosts[Mutator];
+  assert (beg_off >= leftmost_empty(Mutator),
+          "free empty regions before the leftmost: " SIZE_FORMAT ", bound " SIZE_FORMAT, beg_off, leftmost_empty(Mutator));
+  assert (end_off <= rightmost_empty(Mutator),
+          "free empty regions past the rightmost: " SIZE_FORMAT ", bound " SIZE_FORMAT,  end_off, rightmost_empty(Mutator));
+
+  // Performance invariants. Failing these would not break the free set, but performance would suffer.
+  assert (leftmost(Collector) <= _max, "leftmost in bounds: "  SIZE_FORMAT " < " SIZE_FORMAT, leftmost(Collector),  _max);
+  assert (rightmost(Collector) < _max, "rightmost in bounds: "  SIZE_FORMAT " < " SIZE_FORMAT, rightmost(Collector),  _max);
+
+  assert (leftmost(Collector) == _max || in_free_set(leftmost(Collector), Collector),
+          "leftmost region should be free: " SIZE_FORMAT,  leftmost(Collector));
+  assert (leftmost(Collector) == _max || in_free_set(rightmost(Collector), Collector),
+          "rightmost region should be free: " SIZE_FORMAT, rightmost(Collector));
+
+  // If Collector set is empty, leftmosts will both equal max, rightmosts will both equal zero.  Likewise for empty region sets.
+  beg_off = leftmosts[Collector];
+  end_off = rightmosts[Collector];
+  assert (beg_off >= leftmost(Collector),
+          "free regions before the leftmost: " SIZE_FORMAT ", bound " SIZE_FORMAT, beg_off, leftmost(Collector));
+  assert (end_off <= rightmost(Collector),
+          "free regions past the rightmost: " SIZE_FORMAT ", bound " SIZE_FORMAT,  end_off, rightmost(Collector));
+
+  beg_off = empty_leftmosts[Collector];
+  end_off = empty_rightmosts[Collector];
+  assert (beg_off >= leftmost_empty(Collector),
+          "free empty regions before the leftmost: " SIZE_FORMAT ", bound " SIZE_FORMAT, beg_off, leftmost_empty(Collector));
+  assert (end_off <= rightmost_empty(Collector),
+          "free empty regions past the rightmost: " SIZE_FORMAT ", bound " SIZE_FORMAT,  end_off, rightmost_empty(Collector));
+}
+#endif
+
+ShenandoahFreeSet::ShenandoahFreeSet(ShenandoahHeap* heap, size_t max_regions) :
+  _heap(heap),
+  _free_sets(max_regions, this)
+{
+  clear_internal();
 }
 
 HeapWord* ShenandoahFreeSet::allocate_single(ShenandoahAllocRequest& req, bool& in_new_region) {
+  shenandoah_assert_heaplocked();
```

Comment:
In addition, another precondition for this method appears to be that req.size() <= humongous size threshold. Perhaps that check should also be disposed of here.

Reply:
Added similar documentation here.  Thanks.

Reply:
@kdnilsen ... bumping this up. I see that the documentation was added for precondition, but not the check/assert itself here in the implementation. Commenting to unresolve previously resolved comment.

Reply:
... unless this gets checked somewhere more appropriate in the callee chain. When documented as a precondition, it's often easiest to catch debug if asserted upstream rather than deep downstream where context may be difficult to work backwards to.